File: models/flashcard.py
"""
models/flashcard.py - Data models for the flashcard app
"""
import json
import uuid
import random
from dataclasses import dataclass, field
from datetime import datetime
from enum import Enum
from typing import List, Optional, Dict



class QuestionType(Enum):
    SINGLE_CHOICE = "single_choice"
    MULTIPLE_CHOICE = "multiple_choice"
    UNKNOWN = "unknown"


@dataclass
class Flashcard:
    question: str
    options: List[str]                    # ["A. 255", "B. 128", "C. 1024"]
    correct_answers: List[str]            # ["A"] or ["A", "C"] for multi-answer
    question_type: QuestionType = QuestionType.SINGLE_CHOICE
    image_path: Optional[str] = None
    card_id: str = field(default_factory=lambda: str(uuid.uuid4()))
    notes: str = ""

    def to_dict(self) -> dict:
        return {
            "card_id": self.card_id,
            "question": self.question,
            "options": self.options,
            "correct_answers": self.correct_answers,
            "question_type": self.question_type.value,
            "image_path": self.image_path,
            "notes": self.notes,
        }

    @classmethod
    def from_dict(cls, data: dict) -> "Flashcard":
        return cls(
            card_id=data.get("card_id", str(uuid.uuid4())),
            question=data.get("question", ""),
            options=data.get("options", []),
            correct_answers=data.get("correct_answers", []),
            question_type=QuestionType(data.get("question_type", "unknown")),
            image_path=data.get("image_path"),
            notes=data.get("notes", ""),
        )

    def get_correct_answer_text(self) -> str:
        """Return the full text of correct answers."""
        result = []
        for letter in self.correct_answers:
            for opt in self.options:
                if opt.startswith(f"{letter}.") or opt.startswith(f"{letter})"):
                    result.append(opt)
                    break
            else:
                result.append(letter)
        return " | ".join(result) if result else "Unknown"

    # Safe format separators — unique enough to never appear in real content
    SAFE_TERM_SEP = "{[(DapAn)]}"   # between question and answer
    SAFE_CARD_SEP = "{[(CauHoi)]}"  # between cards (replaces newline)

    def to_quizlet_row(self, format_type: str = "full") -> str:
        """
        Export to Quizlet tab-separated row.
        IMPORTANT: Must NOT use \\n inside term or definition.

        format_type:
          - 'simple' : Question [TAB] Correct answer
          - 'full'   : Question >> A | B | C [TAB] Correct answer (full text)
          - 'compact': Question [A | B | C] [TAB] Answer letter(s)
          - 'safe'   : {[(CauHoi)]}Question >> A / B / C{[(DapAn)]}Answer
                       In Quizlet custom import:
                         Giua thuat ngu va dinh nghia → {[(DapAn)]}
                         Giua cac the → {[(CauHoi)]}
        """
        if format_type == "simple":
            term = self.question
            definition = self.get_correct_answer_text()
            return f"{term}\t{definition}"

        elif format_type == "full":
            opts = " | ".join(self.options)
            term = f"{self.question}  >>  {opts}"
            definition = self.get_correct_answer_text()
            return f"{term}\t{definition}"

        elif format_type == "safe":
            # Each option on its own line, no >> or / separators
            opts_lines = "\n".join(self.options)
            term = f"{self.question}\n{opts_lines}"
            definition = self.get_correct_answer_text()
            return f"{self.SAFE_CARD_SEP}{term}{self.SAFE_TERM_SEP}{definition}"

        else:  # compact
            opts = " | ".join(self.options)
            term = f"{self.question} [{opts}]"
            definition = ", ".join(self.correct_answers)
            return f"{term}\t{definition}"



@dataclass
class Deck:
    name: str
    cards: List[Flashcard] = field(default_factory=list)
    deck_id: str = field(default_factory=lambda: str(uuid.uuid4()))
    created_at: str = field(default_factory=lambda: datetime.now().isoformat())
    source_folder: str = ""
    description: str = ""

    def to_dict(self) -> dict:
        return {
            "deck_id": self.deck_id,
            "name": self.name,
            "created_at": self.created_at,
            "source_folder": self.source_folder,
            "description": self.description,
            "cards": [c.to_dict() for c in self.cards],
        }

    @classmethod
    def from_dict(cls, data: dict) -> "Deck":
        cards = [Flashcard.from_dict(c) for c in data.get("cards", [])]
        return cls(
            deck_id=data.get("deck_id", str(uuid.uuid4())),
            name=data.get("name", "Unnamed Deck"),
            created_at=data.get("created_at", datetime.now().isoformat()),
            source_folder=data.get("source_folder", ""),
            description=data.get("description", ""),
            cards=cards,
        )

    @property
    def card_count(self) -> int:
        return len(self.cards)


@dataclass
class QuizSession:
    """Tracks progress for an in-progress quiz on a deck."""
    deck_id: str
    session_id: str = field(default_factory=lambda: str(uuid.uuid4()))
    question_order: List[int] = field(default_factory=list)  # shuffled card indices
    current_index: int = 0                 # which question we're on
    answers: Dict[str, List[str]] = field(default_factory=dict)  # card_id -> chosen answers
    correct_count: int = 0
    wrong_count: int = 0
    started_at: str = field(default_factory=lambda: datetime.now().isoformat())

    @property
    def is_complete(self) -> bool:
        return self.current_index >= len(self.question_order)

    @property
    def progress_frac(self) -> float:
        if not self.question_order:
            return 0.0
        return self.current_index / len(self.question_order)

    def to_dict(self) -> dict:
        return {
            "session_id": self.session_id,
            "deck_id": self.deck_id,
            "question_order": self.question_order,
            "current_index": self.current_index,
            "answers": self.answers,
            "correct_count": self.correct_count,
            "wrong_count": self.wrong_count,
            "started_at": self.started_at,
        }

    @classmethod
    def from_dict(cls, data: dict) -> "QuizSession":
        return cls(
            session_id=data.get("session_id", str(uuid.uuid4())),
            deck_id=data.get("deck_id", ""),
            question_order=data.get("question_order", []),
            current_index=data.get("current_index", 0),
            answers=data.get("answers", {}),
            correct_count=data.get("correct_count", 0),
            wrong_count=data.get("wrong_count", 0),
            started_at=data.get("started_at", datetime.now().isoformat()),
        )

    @classmethod
    def new_for_deck(cls, deck: "Deck") -> "QuizSession":
        order = list(range(len(deck.cards)))
        random.shuffle(order)
        return cls(deck_id=deck.deck_id, question_order=order)

File: appDotNet/FlashcardAI/Models/Flashcard.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace FlashcardAI.Models;

public enum QuestionType
{
    SingleChoice,
    MultipleChoice,
    Unknown
}

public class Flashcard
{
    public string CardId { get; set; } = Guid.NewGuid().ToString();
    public string Question { get; set; } = "";
    public List<string> Options { get; set; } = new();
    public List<string> CorrectAnswers { get; set; } = new();
    public QuestionType QuestionType { get; set; } = QuestionType.SingleChoice;
    public string? ImagePath { get; set; }
    public string Notes { get; set; } = "";
    public int Status { get; set; } = 0; // 0=Gray(Unseen), 1=Orange(Learning/Wrong), 2=Green(Mastered/Correct)

    // Safe format separators
    public const string SafeTermSep = "{[(DapAn)]}";
    public const string SafeCardSep = "{[(CauHoi)]}";

    public string GetCorrectAnswerText()
    {
        var result = new List<string>();
        foreach (var answer in CorrectAnswers)
        {
            var ansTrim = answer.Trim();
            if (ansTrim.Length == 1 && char.IsLetter(ansTrim[0]))
            {
                bool found = false;
                foreach (var opt in Options)
                {
                    if (opt.StartsWith($"{ansTrim}.") || opt.StartsWith($"{ansTrim})"))
                    {
                        result.Add(opt);
                        found = true;
                        break;
                    }
                }
                if (!found) result.Add(answer);
            }
            else
            {
                result.Add(answer);
            }
        }
        return result.Count > 0 ? string.Join(" | ", result) : "Unknown";
    }

    public string ToQuizletRow(string formatType = "full")
    {
        switch (formatType)
        {
            case "simple":
                return $"{Question}\t{GetCorrectAnswerText()}";
            case "full":
            {
                var opts = string.Join(" | ", Options);
                return $"{Question}  >>  {opts}\t{GetCorrectAnswerText()}";
            }
            case "safe":
            {
                var optsLines = string.Join("\n", Options);
                var term = $"{Question}\n{optsLines}";
                return $"{SafeCardSep}{term}{SafeTermSep}{GetCorrectAnswerText()}";
            }
            default: // compact
            {
                var opts = string.Join(" | ", Options);
                var term = $"{Question} [{opts}]";
                return $"{term}\t{string.Join(", ", CorrectAnswers)}";
            }
        }
    }

    public Dictionary<string, object?> ToDict()
    {
        return new Dictionary<string, object?>
        {
            ["card_id"] = CardId,
            ["question"] = Question,
            ["options"] = Options,
            ["correct_answers"] = CorrectAnswers,
            ["question_type"] = QuestionTypeToString(QuestionType),
            ["image_path"] = ImagePath,
            ["notes"] = Notes,
            ["status"] = Status,
        };
    }

    public static Flashcard FromDict(Dictionary<string, JsonElement> data)
    {
        return new Flashcard
        {
            CardId = data.GetStringOr("card_id", Guid.NewGuid().ToString()),
            Question = data.GetStringOr("question", ""),
            Options = data.GetStringListOr("options"),
            CorrectAnswers = data.GetStringListOr("correct_answers"),
            QuestionType = ParseQuestionType(data.GetStringOr("question_type", "unknown")),
            ImagePath = data.TryGetValue("image_path", out var ip) && ip.ValueKind == JsonValueKind.String ? ip.GetString() : null,
            Notes = data.GetStringOr("notes", ""),
            Status = data.TryGetValue("status", out var s) && s.ValueKind == JsonValueKind.Number ? s.GetInt32() : 0,
        };
    }

    public static string QuestionTypeToString(QuestionType qt) => qt switch
    {
        QuestionType.SingleChoice => "single_choice",
        QuestionType.MultipleChoice => "multiple_choice",
        _ => "unknown"
    };

    public static QuestionType ParseQuestionType(string s) => s switch
    {
        "single_choice" => QuestionType.SingleChoice,
        "multiple_choice" => QuestionType.MultipleChoice,
        _ => QuestionType.Unknown
    };
}

public class Deck
{
    public string DeckId { get; set; } = Guid.NewGuid().ToString();
    public string Name { get; set; } = "";
    public string CreatedAt { get; set; } = DateTime.Now.ToString("o");
    public string UpdatedAt { get; set; } = DateTime.Now.ToString("o");
    public string SourceFolder { get; set; } = "";
    public string Description { get; set; } = "";
    public List<Flashcard> Cards { get; set; } = new();

    public int CardCount => Cards.Count;

    public Dictionary<string, object?> ToDict()
    {
        return new Dictionary<string, object?>
        {
            ["deck_id"] = DeckId,
            ["name"] = Name,
            ["created_at"] = CreatedAt,
            ["updated_at"] = UpdatedAt,
            ["source_folder"] = SourceFolder,
            ["description"] = Description,
            ["cards"] = Cards.Select(c => c.ToDict()).ToList(),
        };
    }

    public static Deck FromDict(Dictionary<string, JsonElement> data)
    {
        var cards = new List<Flashcard>();
        if (data.TryGetValue("cards", out var cardsEl) && cardsEl.ValueKind == JsonValueKind.Array)
        {
            foreach (var cardEl in cardsEl.EnumerateArray())
            {
                var dict = JsonSerializer.Deserialize<Dictionary<string, JsonElement>>(cardEl.GetRawText())!;
                cards.Add(Flashcard.FromDict(dict));
            }
        }
        return new Deck
        {
            DeckId = data.GetStringOr("deck_id", Guid.NewGuid().ToString()),
            Name = data.GetStringOr("name", "Unnamed Deck"),
            CreatedAt = data.GetStringOr("created_at", DateTime.Now.ToString("o")),
            UpdatedAt = data.GetStringOr("updated_at", DateTime.Now.ToString("o")),
            SourceFolder = data.GetStringOr("source_folder", ""),
            Description = data.GetStringOr("description", ""),
            Cards = cards,
        };
    }
}

public class QuizSession
{
    public string SessionId { get; set; } = Guid.NewGuid().ToString();
    public string DeckId { get; set; } = "";
    public List<int> QuestionOrder { get; set; } = new();
    public int CurrentIndex { get; set; } = 0;
    public Dictionary<string, List<string>> Answers { get; set; } = new();
    public int CorrectCount { get; set; } = 0;
    public int WrongCount { get; set; } = 0;
    public string StartedAt { get; set; } = DateTime.Now.ToString("o");
    public string UpdatedAt { get; set; } = DateTime.Now.ToString("o");

    public bool IsComplete => CurrentIndex >= QuestionOrder.Count;

    public double ProgressFrac => QuestionOrder.Count == 0 ? 0.0 : (double)CurrentIndex / QuestionOrder.Count;

    public Dictionary<string, object?> ToDict()
    {
        return new Dictionary<string, object?>
        {
            ["session_id"] = SessionId,
            ["deck_id"] = DeckId,
            ["question_order"] = QuestionOrder,
            ["current_index"] = CurrentIndex,
            ["answers"] = Answers,
            ["correct_count"] = CorrectCount,
            ["wrong_count"] = WrongCount,
            ["started_at"] = StartedAt,
            ["updated_at"] = UpdatedAt,
        };
    }

    public static QuizSession FromDict(Dictionary<string, JsonElement> data)
    {
        var answers = new Dictionary<string, List<string>>();
        if (data.TryGetValue("answers", out var ansEl) && ansEl.ValueKind == JsonValueKind.Object)
        {
            foreach (var prop in ansEl.EnumerateObject())
            {
                var list = new List<string>();
                if (prop.Value.ValueKind == JsonValueKind.Array)
                    foreach (var item in prop.Value.EnumerateArray())
                        list.Add(item.GetString() ?? "");
                answers[prop.Name] = list;
            }
        }

        var questionOrder = new List<int>();
        if (data.TryGetValue("question_order", out var qo) && qo.ValueKind == JsonValueKind.Array)
            foreach (var item in qo.EnumerateArray())
                questionOrder.Add(item.GetInt32());

        return new QuizSession
        {
            SessionId = data.GetStringOr("session_id", Guid.NewGuid().ToString()),
            DeckId = data.GetStringOr("deck_id", ""),
            QuestionOrder = questionOrder,
            CurrentIndex = data.TryGetValue("current_index", out var ci) && ci.ValueKind == JsonValueKind.Number ? ci.GetInt32() : 0,
            Answers = answers,
            CorrectCount = data.TryGetValue("correct_count", out var cc) && cc.ValueKind == JsonValueKind.Number ? cc.GetInt32() : 0,
            WrongCount = data.TryGetValue("wrong_count", out var wc) && wc.ValueKind == JsonValueKind.Number ? wc.GetInt32() : 0,
            StartedAt = data.GetStringOr("started_at", DateTime.Now.ToString("o")),
            UpdatedAt = data.GetStringOr("updated_at", DateTime.Now.ToString("o")),
        };
    }

    public static QuizSession NewForDeck(Deck deck)
    {
        var order = Enumerable.Range(0, deck.Cards.Count).ToList();
        var rng = new Random();
        for (int i = order.Count - 1; i > 0; i--)
        {
            int j = rng.Next(i + 1);
            (order[i], order[j]) = (order[j], order[i]);
        }
        return new QuizSession { DeckId = deck.DeckId, QuestionOrder = order };
    }
}

// Extension methods for JSON dictionary parsing
public static class JsonDictExtensions
{
    public static string GetStringOr(this Dictionary<string, JsonElement> dict, string key, string defaultValue = "")
    {
        if (dict.TryGetValue(key, out var el) && el.ValueKind == JsonValueKind.String)
            return el.GetString() ?? defaultValue;
        return defaultValue;
    }

    public static List<string> GetStringListOr(this Dictionary<string, JsonElement> dict, string key)
    {
        var list = new List<string>();
        if (dict.TryGetValue(key, out var el) && el.ValueKind == JsonValueKind.Array)
            foreach (var item in el.EnumerateArray())
                list.Add(item.GetString() ?? "");
        return list;
    }
}
